File: week-3/best-time-to-buy-and-sell-stock-recursive-memoized/best-time-to-buy-and-sell-stock-recursive-memoized.py
class Solution:
    def __init__(self):
        self.memo = {}
        
    def maxProfit(self, prices: List[int], min_buy = float('inf')) -> int:
        state = str(prices) + '~~' + str(min_buy)
        if len(prices) == 0:
            self.memo[state] = 0
        else:
            if prices[0] < min_buy:
                min_buy = prices[0]
        
            self.memo[state] = max(self.maxProfit(prices[1:], min_buy), prices[0] - min_buy)
        return self.memo[state]

File: week-3/best-time-to-buy-and-sell-stock-recursive-memoized/best-time-to-buy-and-sell-stock-recursive-memoized.js
/**
 * @param {number[]} prices
 * @return {number}
 */
const memo = {};

var maxProfit = function(prices, min_buy = Infinity) {
    const state = prices.toString() + '~~' + min_buy;
    if (prices.length === 0) memo[state] = 0;
    else {
        if (prices[0] < min_buy) min_buy = prices[0];
        memo[state] = Math.max(maxProfit(prices.slice(1), min_buy), prices[0] - min_buy);
    }
    
    return memo[state];
};
